causync.py

```diff
@@ -0,0 +1,291 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+""" Rsync wrapper for CausalityGroup """
+
+import logging
+import os
+from shutil import rmtree
+import sys
+from argparse import ArgumentParser
+import subprocess
+from datetime import datetime
+
+import config
+
+class CauSync(object):
+    """ CauSync object for Sync-related functions. """
+    def __init__(self,
+                 cs_config,
+                 src,
+                 dst,
+                 checkonly=False,
+                 cleanup=False,
+                 incremental=False,
+                 incremental_basedir=None,
+                 quiet=False,
+                 selfname="causync.py"):
+
+        self.config = cs_config
+        self.name = selfname
+        self.pid = None
+
+        self.src = src
+        self.src_abs = os.path.abspath(self.src)
+        self.src_basedir = self.src_abs.rstrip('/').split('/')[-1]
+        self.src_parent_dir = self.src_abs.rstrip('/').rsplit('/', 1)[0]
+        self.src_abs = os.path.abspath(src)
+        self.src_lockfile_path = "{}/{}.lock".format(self.src_parent_dir, self.src_basedir)
+
+        self.dst = dst
+        self.dst_abs = os.path.abspath(dst)
+
+        self.cleanup = cleanup
+        self.checkonly = checkonly
+        self.incremental = incremental
+        self.incremental_basedir = os.path.abspath(incremental_basedir) if incremental else None
+        self.quiet = quiet
+
+        self.logger = self.get_logger()
+
+    def run(self):
+        """ Main run function. """
+        self.logger.info("started with PID {}".format(self.pid))
+
+        if self.cleanup:
+            self.run_cleanup(self.incremental)
+        else:
+            if self.lockfile_exists() or self.is_running():
+                self.logger.info("causync is already running on {}".format(self.src_abs))
+            else:
+                self.logger.info("causync is not already running on {}".format(self.src_abs))
+
+            if not self.checkonly:
+                self.run_sync()
+
+    def run_sync(self):
+        """ This function is executed when the task argument is 'sync'. """
+
+        curdate = datetime.now().strftime(config.DATE_FORMAT)
+        extra_flags = ""
+
+        if not self.incremental:
+            dst = os.path.abspath("{dst}/{src}_{d}".format(dst=self.dst,
+                                                           src=self.src_basedir,
+                                                           d=curdate))
+        else:
+            inc_basedir_dirname = self.incremental_basedir.rstrip('/').rsplit('/', 1)[1]
+            dst = os.path.abspath("{dst}/{bd}_inc_{d}".format(dst=self.dst,
+                                                              bd=inc_basedir_dirname,
+                                                              d=curdate))
+
+        if self.incremental and self.incremental_basedir:
+            extra_flags = "--link-dest={}".format(self.incremental_basedir)
```

Comment:
It would be better to use multiple directories for previous backups, so we get incrementality even if a few days are missing.